MoreLinq/Memoize.cs

```diff
@@ -93,19 +96,22 @@ public IEnumerator<T> GetEnumerator()
             {
                 while (true)
                 {
-                    if (index < cache.Count)
+                    lock (locker)
```

Comment:
This puts a lock around the writing but it's [not synchronised with the reading](https://github.com/morelinq/MoreLINQ/pull/211/commits/7709006c7de64837bc756c6d5ab213fbc741e6d6#diff-0cc36bebb7111c5eb462ec0ad60bdd27R118).